### PR TITLE
IRGen: Fix partial_apply of instantiated generic return values

### PR DESCRIPTION
**Pull request by aschwaighofer**
We might have to reabstract the return type.

SR-4253

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test

File: test/IRGen/partial_apply.sil [hunk 1]
```diff
@@ -437,3 +437,63 @@ bb0(%0 : $Int):
   %result = tuple ()
   return %result : $()
 }
+
+
+// Crash on partial apply of a generic enum.
+enum GenericEnum<T> {
+	case X(String)
```
Comment (slavapestov) [inline test/IRGen/partial_apply.sil]: Whitespace issue?

File: test/IRGen/partial_apply.sil [hunk 2]
```diff
@@ -437,3 +437,63 @@ bb0(%0 : $Int):
...
+  %fn = function_ref @generic_indirect_return :$@convention(thin) <T> (Int) -> @owned GenericEnum<T>
+  %pa = partial_apply %fn<Int> (%0) : $@convention(thin) <T> (Int) -> @owned GenericEnum<T>
+  return %pa : $@callee_owned () -> @owned GenericEnum<Int>
+
+}
+
+// Crash on partial apply of a generic enum.
+enum GenericEnum2<T> {
+	case X(String)
+  case Y(T)
```
Comment (slavapestov) [inline test/IRGen/partial_apply.sil]: Again?

File: test/IRGen/partial_apply.sil (hunk 1 above)
Comment (aschwaighofer) [inline test/IRGen/partial_apply.sil]: tabs and spaces ...

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test

File: test/IRGen/partial_apply.sil (hunk 1 above)
Comment (slavapestov) [inline test/IRGen/partial_apply.sil]: In Factor we solved this problem by making tabs an error in the lexer. I don't think that would sit well with swift-evolution though...

**Comment by RLovelett** [comment]
Is this going to be applied to the Swift 3.1 branch as well?

**Pull request merged**